x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/table/RowInTable.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.compute.aggregation.table;
+
+import org.elasticsearch.common.unit.ByteSizeValue;
+import org.elasticsearch.compute.data.Block;
+import org.elasticsearch.compute.data.BlockFactory;
+import org.elasticsearch.compute.data.ElementType;
+import org.elasticsearch.compute.data.IntBlock;
+import org.elasticsearch.compute.data.IntVector;
+import org.elasticsearch.compute.data.Page;
+import org.elasticsearch.core.Releasable;
+import org.elasticsearch.core.ReleasableIterator;
+
+/**
+ * Consumes {@link Page}s and looks up each row in a pre-built table, and returns the
+ * offsets of each row in the table.
+ */
+public abstract sealed class RowInTable implements Releasable permits EmptyRowInTable, AscendingSequenceRowInTable, BlockHashRowInTable {
+    /**
+     * Lookup the values in the {@link Page} and, for each row, return the offset in the
+     * table that was provided when building the lookup.
+     * <p>
+     *     The returned {@link ReleasableIterator} may retain a reference to {@link Block}s
+     *     inside the {@link Page}. Close it to release those references.
+     * </p>
+     */
+    public abstract ReleasableIterator<IntBlock> lookup(Page page, ByteSizeValue targetBlockSize);
+
+    @Override
+    public abstract String toString();
+
+    public static RowInTable build(BlockFactory blockFactory, Block[] keys) {
+        int positions = keys[0].getPositionCount();
+        for (int k = 0; k < keys.length; k++) {
+            if (positions != keys[k].getPositionCount()) {
+                // TODO double check these errors over REST once we have LOOKUP
+                throw new IllegalArgumentException(
+                    "keys must have the same number of positions but [" + positions + "] != [" + keys[k].getPositionCount() + "]"
+                );
+            }
+            for (int p = 0; p < keys[k].getPositionCount(); p++) {
```

Comment:
Maybe a quick check with `Block#mayHaveMultivaluedFields()`, then double-check every position.

Reply:
:+1:. no need to check if it can't have it.